README.md
[![Custom badge](https://img.shields.io/endpoint?color=%2374b9ff&url=https%3A%2F%2Fraw.githubusercontent.com%2unfoldingWord%2Fcontent-validation%2Fmaster%2Fcoverage%2Fshields.json)]()
[![Install, Build & Run Cypress](https://github.com/unfoldingWord/uw-content-validation/workflows/Install,%20Build%20&%20Run%20Cypress/badge.svg)]()

# uW Content/Resource Validation functions

GH Pages: [[https://unfoldingword.github.io/uw-content-validation/]]

This repository contains JavaScript functions for validating/checking for errors in text that is passed to the functions. This text might be a line in a file (especially a TSV file when a line contains a number of distinct fields), or the entire text of a file that’s perhaps open in an editor in the enclosing app.

The basic functions return an object containing two lists:

1. successList: a list of strings giving an overview of what checks have been made,
1. noticeList: a list of objects with fields that can be filtered, sorted, combined, and then displayed as error or warning messages.

Note that the object may also contain other relevant fields such as `checkedFileCount`, `checkedFilenames`, `checkedFilenameExtensions`, `checkedFilesizes`, `checkedRepoNames`, and `elapsedSeconds`.

There are three sample notice processing functions that show how to:

1. Divide the noticeList into a list of errors and a list of warnings,
1. Divide the noticeList into a list of severe, medium, and low priority warnings,
1. Convert the noticeList into a list of warnings sorted by priority,

In addition, there are Styleguidist pages viewable at [[https://unfoldingword.github.io/uw-content-validation/]] which show how these core functions may be used, effectively producing a primitive app that checks Door43.org files, repositories (repos), book packages, etc. as well as demonstrating the basic functions.

## The Stack

- Javascript for primary functions.
- React (functional components & hooks) for demonstration applets.
- MaterialUI for UI/UX baseline design components.
- Styleguidist for Playground Documentation.
- Yarn for dependencies, publishing, and deploying.
- Github + NPM + Github Pages for Hosting.
- Cypress for testing

## Design Philosophy

This code is designed to thoroughly check various types of Bible-related content data files. This includes:

1. [Unified Standard Format Marker](ubsicap.github.io/usfm/) (USFM) Bible content files, including original language Bibles and Bible translations aligned by word/phrase to the original words/phrases
1. Legacy Translation Notes (TN) tables in Tab-Separated Values (9-column TSV) files
1. New tables in Tab-Separated Values (TSV) files (for TWL, TN2 and TQ, SN, and SQ)
1. Markdown files (for TA, TW, and TQ1; also markdown fields in TSV files)
1. Plain-text files
1. Metadata (manifest) YAML files

Note: There is also a separate function for checking individual TSV lines (e.g., TN, TN2, TQ) which is intended to be able to provide immediate user feedback if built into a TSV editor.

The top-level checking demonstrations return:

1. A list of things that were checked (successList)
1. Typically a list of (higher-priority) errors and a list of (lower-priority) warnings, but other formats for display of messages are also demonstrated.

Note that it’s quite possible for the package to give multiple different notices for a single mistake. For example, a punctuation abnormality in a quoted text might advise about the bad punctuation as well as advising about not being able to match the quote. The package errs on the side that having additional warnings is better than missing a warning (and then later being confused about why a data file won’t load). These consequential warnings mean that, if possible, it’s good to rerun the checks from time to time on your latest files as you're fixing multiple errors.

### Notice Objects (in noticeList)

However, the lower-level checking functions provide only the list of success message strings and one list of `notices` (i.e., warnings/errors combined) typically consisting of an object with some or all of the following fields (as available/relevant):

There are two compulsory fields in all of these notice objects:

1. `priority`: A notice priority number in the range 1-1,000. Each different type of warning/error has a unique number (but not each instance of those warnings/errors). By default, notice priority numbers 700 and over are considered `errors` and 0-699 are considered `warnings`, but in truth, that’s rather arbitrary.
1. `message`: The actual general descriptive text of the notice (starts with a capital letter)

All of the following fields may be missing or undefined, i.e., they’re all optional:

1. `details`: More helpful details about the notice (if applicable; doesn’t start with a capital letter)
1. `repoCode`: brief repository code (if available), e.g., 'UHB', 'LT', 'ST', 'TN', 'TQ', forthcoming 'TN2', old 'TQ1', etc.
1. `repoName`: Door43 repository name (if available), e.g., 'en_ta', 'hi_tw'
1. `filename`: filename string (if available)
1. `bookID`: The 3-character UPPERCASE [book identifier](http://ubsicap.github.io/usfm/identification/books.html) or [OBS](https://www.openbiblestories.org/) (if relevant)
1. `C`: The chapter number or OBS story number (if relevant)
1. `V`: The verse number or OBS frame number (if relevant)
1. `rowID`: 4-character ID field for TSV row (if relevant)
1. `lineNumber`: A one-based line number in the file (if available)
1. `fieldName`: name of TSV field (if relevant)
1. `characterIndex`: A **zero-based** integer character index which indicates the position of the error in the given text (line or field) (if available)
1. `excerpt`: An excerpt (if available) from the checked text which indicates the area containing the problem. Where helpful, some character substitutions have already been made, for example, if the notice is about spaces, it is generally helpful to display spaces as a visible character in an attempt to best highlight the issue to the user. (The length of the excerpt defaults to ten characters, but is settable as an option.)
1. `location`: A string indicating the context of the notice, e.g., "in line 17 of 'someBook.usfm'". (Still not completely sure what should be left in this string now that we have added optional `repoName`, `filename`, `rowID`, `lineNumber`, `fieldName` fields.)
1. `extra`: for a check that looks in multiple repos, this contains extra identifying information (typically the `repoCode`) to help the user determine what resource/repo/file that the notice applies to (which, in the demos, is then often prepended to the `message`).

Keeping our notices in this format, rather than the simplicity of just saving an array of single strings, allows the above *notice components* to be processed at a higher level, e.g., to allow user-controlled filtering, sorting, etc. For example, in a rush it might be good to display the highest priority messages first, and fix a few of those. On the other hand, if working systematically through, it might be good to sort by filename and line-number so that warnings about the same part of a file can be viewed together irrespective of their different priority numbers.

The default in the demos is to funnel all the raw notices through the supplied `processNoticesToErrorsWarnings` function (in demos/notice-processing-functions.fs) which does the following:

1. Removes excess repeated errors. For example, if there’s a systematic error in a file, say with unneeded leading spaces in every field, rather than returning with hundreds of errors, only the first several errors will be returned, followed by an "errors suppressed" message. (The number of each error displayed is settable as an option—zero means display all errors with no suppression.)
1. Separates notices into error and warning lists based on the priority number. (The switch-over point is settable as an option.)
1. Optionally drops the lowest priority notices and/or certain given notice types (by priority number).

There is a second version of the function which splits into `Severe`, `Medium`, and `Low` priority lists instead. And a third version that leaves them as notices, but allows for a Bright red...Dull red colour gradient instead.

However, the user is, of course, free to create their own alternative version of these functions. This is possibly also the place to consider localisation of all the notices into different interface languages???

## User-settable Options

### Checking Options

There is provision for checking to be altered and/or sped-up when the calling app sets some or all of the following fields in `checkingOptions`:

- `disableAllLinkFetchingFlag`: a boolean (true/false) which if set to true, stops the package from fetching (hence checking) links, e.g., when a translation note refers to Translation Academy it won’t check that the TA article actually exists, and also stops the checking of any extra files like LICENSE.md—this gives a dramatic speed-up to many checks (but, of course, it means that the data might still contain quite major errors)

CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/unfoldingWord/uw-content-validation/tree/HEAD)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/3.1.0...HEAD)

**Implemented enhancements:**

- Squiggly brackets are now allowed/used in the ULT [\#232](https://github.com/unfoldingWord/uw-content-validation/issues/232)
- Check x-occurrences in aligned materials [\#231](https://github.com/unfoldingWord/uw-content-validation/issues/231)
- Handle alignment across verse ranges [\#228](https://github.com/unfoldingWord/uw-content-validation/issues/228)
- Check for text numbers without thousands dividers [\#193](https://github.com/unfoldingWord/uw-content-validation/issues/193)

**Closed issues:**

- Check if we catch a space after a comma in a number [\#201](https://github.com/unfoldingWord/uw-content-validation/issues/201)
- Cache OrigL verse lookups [\#196](https://github.com/unfoldingWord/uw-content-validation/issues/196)

## [3.1.0](https://github.com/unfoldingWord/uw-content-validation/tree/3.1.0) (2021-12-10)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/2.2.5...3.1.0)

**Implemented enhancements:**

- Detect missing footnote closer [\#207](https://github.com/unfoldingWord/uw-content-validation/issues/207)

**Fixed bugs:**

- Searches Door43-Catalog for new repos [\#208](https://github.com/unfoldingWord/uw-content-validation/issues/208)

**Closed issues:**

- Content Validation v3 [\#226](https://github.com/unfoldingWord/uw-content-validation/issues/226)
- Enhancement: Do existence check for linked articles instead of downloads [\#222](https://github.com/unfoldingWord/uw-content-validation/issues/222)
- TWL CV Results for Book of Acts [\#221](https://github.com/unfoldingWord/uw-content-validation/issues/221)
- Problems with TWL [\#220](https://github.com/unfoldingWord/uw-content-validation/issues/220)
- Suggestion: use resource specific API wrappers instead of repoCodes [\#219](https://github.com/unfoldingWord/uw-content-validation/issues/219)
- Possible false notices for the tN resource for book of Acts [\#218](https://github.com/unfoldingWord/uw-content-validation/issues/218)
- Try to make parameter order more consistent at next version change [\#187](https://github.com/unfoldingWord/uw-content-validation/issues/187)

**Merged pull requests:**

- v3.1 with various small fixes and improvements [\#229](https://github.com/unfoldingWord/uw-content-validation/pull/229) ([RobH123](https://github.com/RobH123))
- v3.0.0 [\#217](https://github.com/unfoldingWord/uw-content-validation/pull/217) ([RobH123](https://github.com/RobH123))

## [2.2.5](https://github.com/unfoldingWord/uw-content-validation/tree/2.2.5) (2021-09-28)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/2.2.2...2.2.5)

**Merged pull requests:**

- Version 2.2.5 [\#216](https://github.com/unfoldingWord/uw-content-validation/pull/216) ([RobH123](https://github.com/RobH123))

## [2.2.2](https://github.com/unfoldingWord/uw-content-validation/tree/2.2.2) (2021-09-03)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/2.2.1...2.2.2)

**Merged pull requests:**

- Fix bug with repoCode for linked articles; add test for newLine in TN single line field that crashed tC build [\#206](https://github.com/unfoldingWord/uw-content-validation/pull/206) ([RobH123](https://github.com/RobH123))

## [2.2.1](https://github.com/unfoldingWord/uw-content-validation/tree/2.2.1) (2021-09-01)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/2.2.0...2.2.1)

**Implemented enhancements:**

- Check strong and x-strong fields in BP check, etc. [\#192](https://github.com/unfoldingWord/uw-content-validation/issues/192)
- Add check for en\_uhal [\#191](https://github.com/unfoldingWord/uw-content-validation/issues/191)
- Add check for en\_ugl [\#190](https://github.com/unfoldingWord/uw-content-validation/issues/190)

**Merged pull requests:**

- Version 2.2.1 [\#205](https://github.com/unfoldingWord/uw-content-validation/pull/205) ([RobH123](https://github.com/RobH123))

## [2.2.0](https://github.com/unfoldingWord/uw-content-validation/tree/2.2.0) (2021-08-12)

[Full Changelog](https://github.com/unfoldingWord/uw-content-validation/compare/2.1.7...2.2.0)

**Closed issues:**

- Investigate EST 2:13 occurrence note issue [\#179](https://github.com/unfoldingWord/uw-content-validation/issues/179)

**Merged pull requests:**

- Version 2.2 [\#200](https://github.com/unfoldingWord/uw-content-validation/pull/200) ([RobH123](https://github.com/RobH123))


src/core/books/books.js
export const BibleBookData = {
  'GEN': { 'title': 'Genesis', 'usfm': '01-GEN', 'testament': 'old', 'verseCount': 1533, 'verseList': [31, 25, 24, 26, 32, 22, 24, 22, 29, 32, 32, 20, 18, 24, 21, 16, 27, 33, 38, 18, 34, 24, 20, 67, 34, 35, 46, 22, 35, 43, 55, 32, 20, 31, 29, 43, 36, 30, 23, 23, 57, 38, 34, 34, 28, 34, 31, 22, 33, 26] },
  'EXO': { 'title': 'Exodus', 'usfm': '02-EXO', 'testament': 'old', 'verseCount': 1213, 'verseList': [22, 25, 22, 31, 23, 30, 25, 32, 35, 29, 10, 51, 22, 31, 27, 36, 16, 27, 25, 26, 36, 31, 33, 18, 40, 37, 21, 43, 46, 38, 18, 35, 23, 35, 35, 38, 29, 31, 43, 38] },
  'LEV': { 'title': 'Leviticus', 'usfm': '03-LEV', 'testament': 'old', 'verseCount': 859, 'verseList': [17, 16, 17, 35, 19, 30, 38, 36, 24, 20, 47, 8, 59, 57, 33, 34, 16, 30, 37, 27, 24, 33, 44, 23, 55, 46, 34] },
  'NUM': { 'title': 'Numbers', 'usfm': '04-NUM', 'testament': 'old', 'verseCount': 1288, 'verseList': [54, 34, 51, 49, 31, 27, 89, 26, 23, 36, 35, 16, 33, 45, 41, 50, 13, 32, 22, 29, 35, 41, 30, 25, 18, 65, 23, 31, 40, 16, 54, 42, 56, 29, 34, 13] },
  'DEU': { 'title': 'Deuteronomy', 'usfm': '05-DEU', 'testament': 'old', 'verseCount': 959, 'verseList': [46, 37, 29, 49, 33, 25, 26, 20, 29, 22, 32, 32, 18, 29, 23, 22, 20, 22, 21, 20, 23, 30, 25, 22, 19, 19, 26, 68, 29, 20, 30, 52, 29, 12] },
  'JOS': { 'title': 'Joshua', 'usfm': '06-JOS', 'testament': 'old', 'verseCount': 658, 'verseList': [18, 24, 17, 24, 15, 27, 26, 35, 27, 43, 23, 24, 33, 15, 63, 10, 18, 28, 51, 9, 45, 34, 16, 33] },
  'JDG': { 'title': 'Judges', 'usfm': '07-JDG', 'testament': 'old', 'verseCount': 618, 'verseList': [36, 23, 31, 24, 31, 40, 25, 35, 57, 18, 40, 15, 25, 20, 20, 31, 13, 31, 30, 48, 25] },
  'RUT': { 'title': 'Ruth', 'usfm': '08-RUT', 'testament': 'old', 'verseCount': 85, 'verseList': [22, 23, 18, 22] },
  '1SA': { 'title': '1 Samuel', 'usfm': '09-1SA', 'testament': 'old', 'verseCount': 810, 'verseList': [28, 36, 21, 22, 12, 21, 17, 22, 27, 27, 15, 25, 23, 52, 35, 23, 58, 30, 24, 42, 15, 23, 29, 22, 44, 25, 12, 25, 11, 31, 13] },
  '2SA': { 'title': '2 Samuel', 'usfm': '10-2SA', 'testament': 'old', 'verseCount': 695, 'verseList': [27, 32, 39, 12, 25, 23, 29, 18, 13, 19, 27, 31, 39, 33, 37, 23, 29, 33, 43, 26, 22, 51, 39, 25] },
  '1KI': { 'title': '1 Kings', 'usfm': '11-1KI', 'testament': 'old', 'verseCount': 816, 'verseList': [53, 46, 28, 34, 18, 38, 51, 66, 28, 29, 43, 33, 34, 31, 34, 34, 24, 46, 21, 43, 29, 53] },
  '2KI': { 'title': '2 Kings', 'usfm': '12-2KI', 'testament': 'old', 'verseCount': 719, 'verseList': [18, 25, 27, 44, 27, 33, 20, 29, 37, 36, 21, 21, 25, 29, 38, 20, 41, 37, 37, 21, 26, 20, 37, 20, 30] },
  '1CH': { 'title': '1 Chronicles', 'usfm': '13-1CH', 'testament': 'old', 'verseCount': 942, 'verseList': [54, 55, 24, 43, 26, 81, 40, 40, 44, 14, 47, 40, 14, 17, 29, 43, 27, 17, 19, 8, 30, 19, 32, 31, 31, 32, 34, 21, 30] },
  '2CH': { 'title': '2 Chronicles', 'usfm': '14-2CH', 'testament': 'old', 'verseCount': 822, 'verseList': [17, 18, 17, 22, 14, 42, 22, 18, 31, 19, 23, 16, 22, 15, 19, 14, 19, 34, 11, 37, 20, 12, 21, 27, 28, 23, 9, 27, 36, 27, 21, 33, 25, 33, 27, 23] },
  'EZR': { 'title': 'Ezra', 'usfm': '15-EZR', 'testament': 'old', 'verseCount': 280, 'verseList': [11, 70, 13, 24, 17, 22, 28, 36, 15, 44] },
  'NEH': { 'title': 'Nehemiah', 'usfm': '16-NEH', 'testament': 'old', 'verseCount': 406, 'verseList': [11, 20, 32, 23, 19, 19, 73, 18, 38, 39, 36, 47, 31] },
  'EST': { 'title': 'Esther', 'usfm': '17-EST', 'testament': 'old', 'verseCount': 167, 'verseList': [22, 23, 15, 17, 14, 14, 10, 17, 32, 3] },
  'JOB': { 'title': 'Job', 'usfm': '18-JOB', 'testament': 'old', 'verseCount': 1070, 'verseList': [22, 13, 26, 21, 27, 30, 21, 22, 35, 22, 20, 25, 28, 22, 35, 22, 16, 21, 29, 29, 34, 30, 17, 25, 6, 14, 23, 28, 25, 31, 40, 22, 33, 37, 16, 33, 24, 41, 30, 24, 34, 17] },
  'PSA': { 'title': 'Psalms', 'usfm': '19-PSA', 'testament': 'old', 'verseCount': 2461, 'verseList': [6, 12, 8, 8, 12, 10, 17, 9, 20, 18, 7, 8, 6, 7, 5, 11, 15, 50, 14, 9, 13, 31, 6, 10, 22, 12, 14, 9, 11, 12, 24, 11, 22, 22, 28, 12, 40, 22, 13, 17, 13, 11, 5, 26, 17, 11, 9, 14, 20, 23, 19, 9, 6, 7, 23, 13, 11, 11, 17, 12, 8, 12, 11, 10, 13, 20, 7, 35, 36, 5, 24, 20, 28, 23, 10, 12, 20, 72, 13, 19, 16, 8, 18, 12, 13, 17, 7, 18, 52, 17, 16, 15, 5, 23, 11, 13, 12, 9, 9, 5, 8, 28, 22, 35, 45, 48, 43, 13, 31, 7, 10, 10, 9, 8, 18, 19, 2, 29, 176, 7, 8, 9, 4, 8, 5, 6, 5, 6, 8, 8, 3, 18, 3, 3, 21, 26, 9, 8, 24, 13, 10, 7, 12, 15, 21, 10, 20, 14, 9, 6] },
  'PRO': { 'title': 'Proverbs', 'usfm': '20-PRO', 'testament': 'old', 'verseCount': 915, 'verseList': [33, 22, 35, 27, 23, 35, 27, 36, 18, 32, 31, 28, 25, 35, 33, 33, 28, 24, 29, 30, 31, 29, 35, 34, 28, 28, 27, 28, 27, 33, 31] },
  'ECC': { 'title': 'Ecclesiastes', 'usfm': '21-ECC', 'testament': 'old', 'verseCount': 222, 'verseList': [18, 26, 22, 16, 20, 12, 29, 17, 18, 20, 10, 14] },
  'SNG': { 'title': 'Song of Songs', 'usfm': '22-SNG', 'testament': 'old', 'verseCount': 117, 'verseList': [17, 17, 11, 16, 16, 13, 13, 14] },
  'ISA': { 'title': 'Isaiah', 'usfm': '23-ISA', 'testament': 'old', 'verseCount': 1292, 'verseList': [31, 22, 26, 6, 30, 13, 25, 22, 21, 34, 16, 6, 22, 32, 9, 14, 14, 7, 25, 6, 17, 25, 18, 23, 12, 21, 13, 29, 24, 33, 9, 20, 24, 17, 10, 22, 38, 22, 8, 31, 29, 25, 28, 28, 25, 13, 15, 22, 26, 11, 23, 15, 12, 17, 13, 12, 21, 14, 21, 22, 11, 12, 19, 12, 25, 24] },
  'JER': { 'title': 'Jeremiah', 'usfm': '24-JER', 'testament': 'old', 'verseCount': 1364, 'verseList': [19, 37, 25, 31, 31, 30, 34, 22, 26, 25, 23, 17, 27, 22, 21, 21, 27, 23, 15, 18, 14, 30, 40, 10, 38, 24, 22, 17, 32, 24, 40, 44, 26, 22, 19, 32, 21, 28, 18, 16, 18, 22, 13, 30, 5, 28, 7, 47, 39, 46, 64, 34] },
  'LAM': { 'title': 'Lamentations', 'usfm': '25-LAM', 'testament': 'old', 'verseCount': 154, 'verseList': [22, 22, 66, 22, 22] },
  'EZK': { 'title': 'Ezekiel', 'usfm': '26-EZK', 'testament': 'old', 'verseCount': 1273, 'verseList': [28, 10, 27, 17, 17, 14, 27, 18, 11, 22, 25, 28, 23, 23, 8, 63, 24, 32, 14, 49, 32, 31, 49, 27, 17, 21, 36, 26, 21, 26, 18, 32, 33, 31, 15, 38, 28, 23, 29, 49, 26, 20, 27, 31, 25, 24, 23, 35] },
  'DAN': { 'title': 'Daniel', 'usfm': '27-DAN', 'testament': 'old', 'verseCount': 357, 'verseList': [21, 49, 30, 37, 31, 28, 28, 27, 27, 21, 45, 13] },
  'HOS': { 'title': 'Hosea', 'usfm': '28-HOS', 'testament': 'old', 'verseCount': 197, 'verseList': [11, 23, 5, 19, 15, 11, 16, 14, 17, 15, 12, 14, 16, 9] },
  'JOL': { 'title': 'Joel', 'usfm': '29-JOL', 'testament': 'old', 'verseCount': 73, 'verseList': [20, 32, 21] },
  'AMO': { 'title': 'Amos', 'usfm': '30-AMO', 'testament': 'old', 'verseCount': 146, 'verseList': [15, 16, 15, 13, 27, 14, 17, 14, 15] },
  'OBA': { 'title': 'Obadiah', 'usfm': '31-OBA', 'testament': 'old', 'verseCount': 21, 'verseList': [21] },
  'JON': { 'title': 'Jonah', 'usfm': '32-JON', 'testament': 'old', 'verseCount': 48, 'verseList': [17, 10, 10, 11] },
  'MIC': { 'title': 'Micah', 'usfm': '33-MIC', 'testament': 'old', 'verseCount': 105, 'verseList': [16, 13, 12, 13, 15, 16, 20] },
  'NAM': { 'title': 'Nahum', 'usfm': '34-NAM', 'testament': 'old', 'verseCount': 47, 'verseList': [15, 13, 19] },
  'HAB': { 'title': 'Habakkuk', 'usfm': '35-HAB', 'testament': 'old', 'verseCount': 56, 'verseList': [17, 20, 19] },
  'ZEP': { 'title': 'Zephaniah', 'usfm': '36-ZEP', 'testament': 'old', 'verseCount': 53, 'verseList': [18, 15, 20] },
  'HAG': { 'title': 'Haggai', 'usfm': '37-HAG', 'testament': 'old', 'verseCount': 38, 'verseList': [15, 23] },
  'ZEC': { 'title': 'Zechariah', 'usfm': '38-ZEC', 'testament': 'old', 'verseCount': 211, 'verseList': [21, 13, 10, 14, 11, 15, 14, 23, 17, 12, 17, 14, 9, 21] },
  'MAL': { 'title': 'Malachi', 'usfm': '39-MAL', 'testament': 'old', 'verseCount': 55, 'verseList': [14, 17, 18, 6] },
  'MAT': { 'title': 'Matthew', 'usfm': '41-MAT', 'testament': 'new', 'verseCount': 1071, 'verseList': [25, 23, 17, 25, 48, 34, 29, 34, 38, 42, 30, 50, 58, 36, 39, 28, 27, 35, 30, 34, 46, 46, 39, 51, 46, 75, 66, 20] },
  'MRK': { 'title': 'Mark', 'usfm': '42-MRK', 'testament': 'new', 'verseCount': 678, 'verseList': [45, 28, 35, 41, 43, 56, 37, 38, 50, 52, 33, 44, 37, 72, 47, 20] },
  'LUK': { 'title': 'Luke', 'usfm': '43-LUK', 'testament': 'new', 'verseCount': 1151, 'verseList': [80, 52, 38, 44, 39, 49, 50, 56, 62, 42, 54, 59, 35, 35, 32, 31, 37, 43, 48, 47, 38, 71, 56, 53] },
  'JHN': { 'title': 'John', 'usfm': '44-JHN', 'testament': 'new', 'verseCount': 879, 'verseList': [51, 25, 36, 54, 47, 71, 53, 59, 41, 42, 57, 50, 38, 31, 27, 33, 26, 40, 42, 31, 25] },
  'ACT': { 'title': 'Acts', 'usfm': '45-ACT', 'testament': 'new', 'verseCount': 1007, 'verseList': [26, 47, 26, 37, 42, 15, 60, 40, 43, 48, 30, 25, 52, 28, 41, 40, 34, 28, 41, 38, 40, 30, 35, 27, 27, 32, 44, 31] },
  'ROM': { 'title': 'Romans', 'usfm': '46-ROM', 'testament': 'new', 'verseCount': 433, 'verseList': [32, 29, 31, 25, 21, 23, 25, 39, 33, 21, 36, 21, 14, 23, 33, 27] },
  '1CO': { 'title': '1 Corinthians', 'usfm': '47-1CO', 'testament': 'new', 'verseCount': 437, 'verseList': [31, 16, 23, 21, 13, 20, 40, 13, 27, 33, 34, 31, 13, 40, 58, 24] },
  '2CO': { 'title': '2 Corinthians', 'usfm': '48-2CO', 'testament': 'new', 'verseCount': 257, 'verseList': [24, 17, 18, 18, 21, 18, 16, 24, 15, 18, 33, 21, 14] },
  'GAL': { 'title': 'Galatians', 'usfm': '49-GAL', 'testament': 'new', 'verseCount': 149, 'verseList': [24, 21, 29, 31, 26, 18] },
  'EPH': { 'title': 'Ephesians', 'usfm': '50-EPH', 'testament': 'new', 'verseCount': 155, 'verseList': [23, 22, 21, 32, 33, 24] },
  'PHP': { 'title': 'Philippians', 'usfm': '51-PHP', 'testament': 'new', 'verseCount': 104, 'verseList': [30, 30, 21, 23] },
  'COL': { 'title': 'Colossians', 'usfm': '52-COL', 'testament': 'new', 'verseCount': 95, 'verseList': [29, 23, 25, 18] },
  '1TH': { 'title': '1 Thessalonians', 'usfm': '53-1TH', 'testament': 'new', 'verseCount': 89, 'verseList': [10, 20, 13, 18, 28] },
  '2TH': { 'title': '2 Thessalonians', 'usfm': '54-2TH', 'testament': 'new', 'verseCount': 47, 'verseList': [12, 17, 18] },
  '1TI': { 'title': '1 Timothy', 'usfm': '55-1TI', 'testament': 'new', 'verseCount': 113, 'verseList': [20, 15, 16, 16, 25, 21] },
  '2TI': { 'title': '2 Timothy', 'usfm': '56-2TI', 'testament': 'new', 'verseCount': 83, 'verseList': [18, 26, 17, 22] },
  'TIT': { 'title': 'Titus', 'usfm': '57-TIT', 'testament': 'new', 'verseCount': 46, 'verseList': [16, 15, 15] },
  'PHM': { 'title': 'Philemon', 'usfm': '58-PHM', 'testament': 'new', 'verseCount': 25, 'verseList': [25] },
  'HEB': { 'title': 'Hebrews', 'usfm': '59-HEB', 'testament': 'new', 'verseCount': 303, 'verseList': [14, 18, 19, 16, 14, 20, 28, 13, 28, 39, 40, 29, 25] },
  'JAS': { 'title': 'James', 'usfm': '60-JAS', 'testament': 'new', 'verseCount': 108, 'verseList': [27, 26, 18, 17, 20] },
  '1PE': { 'title': '1 Peter', 'usfm': '61-1PE', 'testament': 'new', 'verseCount': 105, 'verseList': [25, 25, 22, 19, 14] },
  '2PE': { 'title': '2 Peter', 'usfm': '62-2PE', 'testament': 'new', 'verseCount': 61, 'verseList': [21, 22, 18] },
  '1JN': { 'title': '1 John', 'usfm': '63-1JN', 'testament': 'new', 'verseCount': 105, 'verseList': [10, 29, 24, 21, 21] },
  '2JN': { 'title': '2 John', 'usfm': '64-2JN', 'testament': 'new', 'verseCount': 13, 'verseList': [13] },
  '3JN': { 'title': '3 John', 'usfm': '65-3JN', 'testament': 'new', 'verseCount': 15, 'verseList': [15] },
  'JUD': { 'title': 'Jude', 'usfm': '66-JUD', 'testament': 'new', 'verseCount': 25, 'verseList': [25] },
  'REV': { 'title': 'Revelation', 'usfm': '67-REV', 'testament': 'new', 'verseCount': 404, 'verseList': [20, 29, 22, 11, 14, 17, 17, 13, 21, 11, 19, 17, 18, 20, 8, 21, 18, 24, 21, 15, 27, 21] }
}

const oftenMissingBCVList = [
  // See https://en.wikipedia.org/wiki/List_of_New_Testament_verses_not_included_in_modern_English_translations
  ['NEH', 7, 68], // ?
  ['MAT', 16, 3],
  ['MAT', 17, 21],
  ['MAT', 18, 11],
  // ['MAT', 20, 16b],
  ['MAT', 23, 14],
  // ['MRK', 6, 11b],
  ['MRK', 7, 16],
  ['MRK', 9, 44], ['MRK', 9, 46],
  ['MRK', 11, 26],
  ['MRK', 15, 28],
  ['MRK', 16, 9], ['MRK', 16, 10], ['MRK', 16, 11], ['MRK', 16, 12], ['MRK', 16, 13], ['MRK', 16, 14],
  ['MRK', 16, 15], ['MRK', 16, 16], ['MRK', 16, 17], ['MRK', 16, 18], ['MRK', 16, 19], ['MRK', 16, 20],
  // ['LUK', 4, 8b],
  // ['LUK', 9, 55], ['LUK', 9, 56],
  ['LUK', 17, 36],
  ['LUK', 22, 43], // ?
  ['LUK', 22, 44], // ?
  ['LUK', 23, 17],
  ['JHN', 5, 3], ['JHN', 5, 4],
  ['JHN', 7, 53], // ?
  ['JHN', 8, 1], // ?
  ['ACT', 8, 37],
  // ['ACT', 9, 5], ['ACT', 9, 6],
  // ['ACT', 13, 42],
  ['ACT', 15, 34],
  // ['ACT', 23, 9b],
  ['ACT', 24, 6], ['ACT', 24, 7], ['ACT', 24, 8],
  ['ACT', 28, 29],
  ['ROM', 16, 24],
  ['2CO', 13, 14], // ?
  ['1JN', 5, 7], ['1JN', 5, 8],
];


const extraBookList = ['FRT', 'BAK'];
export const isValidBookID = (bookId) => {
  return bookId.toUpperCase() in BibleBookData || extraBookList.includes(bookId);
}
export const isOptionalValidBookID = (bookId) => {
  return !bookId || bookId.toUpperCase() in BibleBookData || extraBookList.includes(bookId);
}
export const isExtraBookID = (bookId) => {
  return extraBookList.includes(bookId);
}


export const usfmNumberName = (bookId) => {
  try { return BibleBookData[bookId.toUpperCase()].usfm; }
  catch (err) { throw new Error(`usfmNumberName() given invalid bookId: '${bookId}'`); }
}

export const expectedVersesPerChapterList = (bookId) => {
  // parameterAssert(bookId.toUpperCase() !== 'OBS', `expectedVersesPerChapterList shouldn’t be passed '${bookId}'`);
  let verseList;
  try {
    verseList = BibleBookData[bookId.toUpperCase()].verseList;
  } catch (err) {
    throw new Error(`expectedVersesPerChapterList() given invalid bookId: '${bookId}'`);
  }
  if (verseList === undefined) {
    throw new Error(`expectedVersesPerChapterList(): Invalid bookId: '${bookId}'`);
  }
  return verseList;
};

export const chaptersInBook = (bookId) => {
  return expectedVersesPerChapterList(bookId).length;
};

export const isOneChapterBook = (bookId) => {
  return chaptersInBook(bookId) === 1;
};

export const versesInChapter = (bookId, chapter) => {
  // parameterAssert(bookId.toUpperCase() !== 'OBS', `versesInChapter shouldn’t be passed '${bookId}'`);
  const verses = expectedVersesPerChapterList(bookId)[chapter - 1];
  if (verses === undefined) {
    throw new Error(`versesInChapter(${bookId}) given invalid chapter: ${chapter}`);
  }
  return verses;
};

export const testament = (bookId) => {
  try { return BibleBookData[bookId.toUpperCase()].testament; }
  catch (err) { throw new Error(`testament() given invalid Bible bookId: '${bookId}'`); }
};

export function getEnglishBookName(bookId) {
  try { return BibleBookData[bookId.toUpperCase()].title; }
  catch (err) { throw new Error(`getEnglishBookName() given invalid Bible bookId: '${bookId}'`); }
}

export function isGoodEnglishBookName(givenBookName) {
  // debugLog(`isGoodEnglishBookName(${givenBookName})…`);
  const partialMatches = [];
  const givenBookNameLower = givenBookName.toUpperCase();
  for (const bk in BibleBookData) {
    const thisBookName = BibleBookData[bk].title;
    // debugLog("thisBookName", thisBookName);
    if (thisBookName === givenBookName) return true;
    const thisBookNameLower = thisBookName.toUpperCase();
    if (thisBookNameLower === givenBookNameLower) return 1;
    if (thisBookNameLower.startsWith(givenBookNameLower)) partialMatches.push(thisBookName);
  }
  if (partialMatches.length === 1) return 2; // We got an unambiguous partial match, e.g., Gen for Genesis
  return false;
}

export function isOftenMissing(bookID, C, V) {
  function matchBCV(entry) { return entry[0] === bookID && entry[1] === C && entry[2] === V; }

  return oftenMissingBCVList.find(matchBCV) !== undefined;
}
